src/zkevm_specs/evm/typing.py

```diff
@@ -615,3 +648,141 @@ def _append(
         )
 
         return self
+
+
+class CopyCircuit:
+    rows: List[CopyCircuitRow]
+    pad_rows: List[CopyCircuitRow]
+
+    def __init__(self) -> None:
+        self.rows = []
+        self.pad_rows = [CopyCircuitRow(FQ(1), *[FQ(0)] * 17), CopyCircuitRow(*[FQ(0)] * 18)]
+
+    def table(self) -> Sequence[CopyCircuitRow]:
+        return self.rows + self.pad_rows
+
+    def copy(
+        self,
+        rw_dict: RWDictionary,
+        src_id: IntOrFQ,
+        src_type: CopyDataTypeTag,
+        dst_id: IntOrFQ,
+        dst_type: CopyDataTypeTag,
+        src_addr: IntOrFQ,
+        src_addr_boundary: IntOrFQ,
+        dst_addr: IntOrFQ,
+        copy_length: IntOrFQ,
+        src_data: Mapping[IntOrFQ, Union[IntOrFQ, Tuple[IntOrFQ, IntOrFQ]]],
+        log_id: IntOrFQ = FQ(0),
+    ):
+        new_rows: List[CopyCircuitRow] = []
+        for i in range(int(copy_length)):
+            if int(src_addr + i) < int(src_addr_boundary):
+                is_pad = False
+                assert src_addr in src_data, f"Cannot find data at the offset {src_addr}"
```

Comment:
```suggestion
                assert src_addr+i in src_data, f"Cannot find data at the offset {src_addr+i}"
```
otherwise the real offset is not used in the check

Reply:
fixed in a000bf3